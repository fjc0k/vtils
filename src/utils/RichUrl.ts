import { isPlainObject } from 'lodash-uni'

export interface ParsedRichUrl<TDesc> {
  url: string
  desc?: TDesc
}

export interface ParsedFileRichUrl {
  url: string
  file: File
}

/**
 * 富链接，同普通链接相比，富链接可包含一些描述信息。
 *
 * 结构描述：
 * ```text
 * rich://{"url":"***","desc":"***"}
 * ```
 */
export class RichUrl {
  /**
   * 标识符。
   */
  private static readonly identity = 'rich://'

  /**
   * 检查是否是富链接。
   *
   * @param value 要检查的值
   * @returns 返回检查结果
   * @example
   * ```typescript
   * RichUrl.check('http://www.google.com') // => false
   * ```
   */
  static check(value: any): value is string {
    return (
      typeof value === 'string' &&
      value.substr(0, RichUrl.identity.length) === RichUrl.identity
    )
  }

  /**
   * 创建富链接。
   *
   * @param url 普通链接
   * @param desc 描述信息
   * @returns 返回创建的富链接
   */
  static build(url: string, desc?: any): string {
    return `${RichUrl.identity}${JSON.stringify({ url, desc })}`
  }

  /**
   * 解析富链接。非富链接的会直接将其值作为 url 返回。
   *
   * @param richUrl 富链接
   * @returns 返回解析结果
   */
  static parse<TDesc>(richUrl: string): ParsedRichUrl<TDesc> {
    if (!RichUrl.check(richUrl)) {
      return {
        url: richUrl,
      }
    }
    try {
      return JSON.parse(richUrl.substr(RichUrl.identity.length))
    } catch {
      return {
        url: richUrl,
      }
    }
  }

  /**
   * 转换数据中的富链接。
   *
   * @param data 数据
   * @param callback 回调
   * @returns 返回转换后的数据
   */
  static transform<TData, TDesc>(
    data: TData,
    callback: (
      parsedRichUrl: ParsedRichUrl<TDesc>,
      data: TData,
    ) => Promise<string>,
  ): Promise<TData> {
    return new Promise(resolve => {
      if (Array.isArray(data)) {
        Promise.all(
          (data as any[]).map((value, index) => {
            return RichUrl.transform(value, callback).then(res => {
              ;(data as any[])[index] = res
            })
          }),
        ).then(() => resolve(data))
      } else if (isPlainObject(data)) {
        Promise.all(
          Object.keys(data).map(key => {
            return RichUrl.transform((data as any)[key], callback).then(res => {
              ;(data as any)[key] = res
            })
          }),
        ).then(() => resolve(data))
      } else if (RichUrl.check(data)) {
        callback(RichUrl.parse(data), data).then(resolve as any)
      } else {
        resolve(data)
      }
    })
  }

  /**
   * 将文件转换为文件富链接。
   *
   * @param file 要转换的文件
   * @returns 返回转换后的文件富链接
   */
  static fromFile(file: File): string {
    return RichUrl.build(URL.createObjectURL(file), {
      name: file.name,
      size: file.size,
      type: file.type,
      lastModified: file.lastModified,
    } as Partial<File>)
  }

  /**
   * 将文件富链接转换为文件和普通链接。
   *
   * @param richUrl 要转换的文件富链接
   * @returns 返回转换后的文件和普通链接
   */
  static toFile(
    richUrl: string | ParsedRichUrl<File>,
  ): Promise<ParsedFileRichUrl> {
    return new Promise((resolve, reject) => {
      const { url, desc } =
        typeof richUrl === 'string' ? RichUrl.parse<File>(richUrl) : richUrl
      if (
        !url ||
        !desc ||
        url.substr(0, 5) !== 'blob:' ||
        typeof desc !== 'object'
      ) {
        reject(
          new Error(
            `richUrl 不是一个合法的文件富链接: ${JSON.stringify(richUrl)}`,
          ),
        )
      } else {
        const xhr = new XMLHttpRequest()
        xhr.open('GET', url)
        xhr.responseType = 'blob'
        xhr.onload = () => {
          const file = new File([xhr.response], desc.name, desc)
          resolve({
            url: url,
            file: file,
          })
        }
        xhr.onerror = reject
        xhr.send()
      }
    })
  }

  /**
   * 转换数据中的文件富链接。
   *
   * @param data 数据
   * @param callback 回调
   * @returns 返回转换后的数据
   */
  static transformFile<TData>(
    data: TData,
    callback: (parsedFileRichUrl: ParsedFileRichUrl) => Promise<string>,
  ): Promise<TData> {
    return RichUrl.transform<TData, File>(data, parsedRichUrl => {
      return RichUrl.toFile(parsedRichUrl).then(callback)
    })
  }
}
