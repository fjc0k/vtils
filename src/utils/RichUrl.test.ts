import { RichUrl } from './RichUrl'

describe('RichUrl', () => {
  describe('check', () => {
    test('不是富链接', () => {
      const urls = [
        'http://gg.gg',
        'https://gg.gg',
        'ftp://gg.gg',
        'mailto:gg@gg.gg',
        'rich:gg.gg',
      ]
      for (const url of urls) {
        expect(RichUrl.check(url)).toBeFalse()
      }
    })

    test('是富链接', () => {
      const urls = ['rich://', 'rich://gg.gg', 'rich://{}']
      for (const url of urls) {
        expect(RichUrl.check(url)).toBeTrue()
      }
    })
  })

  describe('build', () => {
    test('无描述', () => {
      expect(RichUrl.build('')).toMatchSnapshot('url为空')
      expect(RichUrl.build('http://gg.gg')).toMatchSnapshot('url不为空')
    })

    test('有描述', () => {
      expect(RichUrl.build('', null)).toMatchSnapshot('描述为 null')
      expect(RichUrl.build('', 1)).toMatchSnapshot('描述为 1')
      expect(RichUrl.build('', [])).toMatchSnapshot('描述为 []')
      expect(RichUrl.build('', { x: 1 })).toMatchSnapshot('描述为 { x: 1 }')
      expect(RichUrl.build('', 'desc')).toMatchSnapshot('描述为 desc')
    })
  })

  describe('parse', () => {
    test('非富链接时返回原始输入作为普通链接', () => {
      const urls = [
        'http://gg.gg',
        'https://gg.gg',
        'ftp://gg.gg',
        'mailto:gg@gg.gg',
        'rich:gg.gg',
        "rich://''",
      ]
      for (const url of urls) {
        expect(RichUrl.parse(url)).toEqual({ url })
      }
    })

    test('富链接时解析正常', () => {
      expect(RichUrl.parse(RichUrl.build(''))).toMatchSnapshot('url为空')
      expect(RichUrl.parse(RichUrl.build('http://gg.gg'))).toMatchSnapshot(
        'url不为空',
      )
      expect(RichUrl.parse(RichUrl.build('', null))).toMatchSnapshot(
        '描述为 null',
      )
      expect(RichUrl.parse(RichUrl.build('', 1))).toMatchSnapshot('描述为 1')
      expect(RichUrl.parse(RichUrl.build('', []))).toMatchSnapshot('描述为 []')
      expect(RichUrl.parse(RichUrl.build('', { x: 1 }))).toMatchSnapshot(
        '描述为 { x: 1 }',
      )
      expect(RichUrl.parse(RichUrl.build('', 'desc'))).toMatchSnapshot(
        '描述为 desc',
      )
    })
  })

  describe('transform', () => {
    test('表现正常', async () => {
      expect(
        await RichUrl.transform(
          {
            id: 1,
            title: 'hello',
            icon: RichUrl.build('icon', { name: 'icon' }),
            users: [
              {
                id: 2,
                avatar: RichUrl.build('2-avatar'),
              },
              {
                id: 3,
                avatar: RichUrl.build('3-avatar', { url: '.' }),
              },
              {
                id: 4,
                avatar: RichUrl.build('', { avatar: '4', id: 4 }),
                albums: [
                  RichUrl.build('http://gg.gg'),
                  RichUrl.build('https://gg.gg'),
                  RichUrl.build('ftp://gg.gg', { ftp: true }),
                ],
              },
            ],
          },
          async ({ url, desc }) => `url: ${url}, desc: ${JSON.stringify(desc)}`,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('fromFile', () => {
    beforeAll(() => {
      window.URL.createObjectURL = () => 'blob:http://localhost/ffsfeeer44'
    })

    afterAll(() => {
      // @ts-ignore
      delete window.URL.createObjectURL
    })

    test('表现正常', () => {
      expect(
        RichUrl.fromFile(
          new File([''], 'h.css', {
            type: 'text/css',
            lastModified: 1574155444339,
          }),
        ),
      ).toMatchSnapshot()
    })
  })

  describe('toFile', () => {
    const rawSend = window.XMLHttpRequest.prototype.send.bind(
      window.XMLHttpRequest.prototype,
    )

    beforeAll(() => {
      window.URL.createObjectURL = () => 'blob:http://localhost/ffsfeeer44'
      window.XMLHttpRequest.prototype.send = function () {
        Object.defineProperty(this, 'response', {
          value: '',
        })
        this.onload?.(this as any)
      }
    })

    afterAll(() => {
      // @ts-ignore
      delete window.URL.createObjectURL
      window.XMLHttpRequest.prototype.send = rawSend
    })

    test('非文件富链接报错', async () => {
      await expect(
        RichUrl.toFile(RichUrl.build('http://gg.gg', { x: 1 })),
      ).rejects.toThrowErrorMatchingSnapshot()
    })

    test('文件富链接正常', async () => {
      const { file, url } = await RichUrl.toFile(
        RichUrl.fromFile(
          new File([''], 'h.css', {
            type: 'text/css',
            lastModified: 1574155444339,
          }),
        ),
      )
      expect({
        url: url,
        name: file.name,
        size: file.size,
        type: file.type,
        lastModified: file.lastModified,
      }).toMatchSnapshot()
    })
  })

  describe('transformFile', () => {
    const rawSend = window.XMLHttpRequest.prototype.send.bind(
      window.XMLHttpRequest.prototype,
    )

    beforeAll(() => {
      window.URL.createObjectURL = () => 'blob:http://localhost/ffsfeeer44'
      window.XMLHttpRequest.prototype.send = function () {
        Object.defineProperty(this, 'response', {
          value: '',
        })
        this.onload?.(this as any)
      }
    })

    afterAll(() => {
      // @ts-ignore
      delete window.URL.createObjectURL
      window.XMLHttpRequest.prototype.send = rawSend
    })

    test('表现正常', async () => {
      expect(
        await RichUrl.transformFile(
          {
            id: 1,
            title: 'hello',
            icon: RichUrl.build('blob://x0', { name: 'icon' }),
            users: [
              {
                id: 2,
                avatar: RichUrl.build('blob://x1', { url: '.' }),
              },
              {
                id: 3,
                avatar: RichUrl.build('blob://x2', { url: '.' }),
              },
              {
                id: 4,
                avatar: RichUrl.fromFile(
                  new File([''], 'h.css', {
                    type: 'text/css',
                    lastModified: 1574155444339,
                  }),
                ),
                albums: [
                  RichUrl.build('blob://x3', { url: '.' }),
                  RichUrl.build('blob://x4', { url: '.' }),
                  RichUrl.build('blob://x5', { ftp: true }),
                ],
              },
            ],
          },
          async ({ url, file }) => `url: ${url}, file: ${file.name}`,
        ),
      ).toMatchSnapshot()
    })
  })
})
