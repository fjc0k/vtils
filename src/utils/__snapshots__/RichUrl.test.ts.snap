// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichUrl build 无描述: url不为空 1`] = `"rich://{\\"url\\":\\"http://gg.gg\\"}"`;

exports[`RichUrl build 无描述: url为空 1`] = `"rich://{\\"url\\":\\"\\"}"`;

exports[`RichUrl build 有描述: 描述为 [] 1`] = `"rich://{\\"url\\":\\"\\",\\"desc\\":[]}"`;

exports[`RichUrl build 有描述: 描述为 { x: 1 } 1`] = `"rich://{\\"url\\":\\"\\",\\"desc\\":{\\"x\\":1}}"`;

exports[`RichUrl build 有描述: 描述为 1 1`] = `"rich://{\\"url\\":\\"\\",\\"desc\\":1}"`;

exports[`RichUrl build 有描述: 描述为 desc 1`] = `"rich://{\\"url\\":\\"\\",\\"desc\\":\\"desc\\"}"`;

exports[`RichUrl build 有描述: 描述为 null 1`] = `"rich://{\\"url\\":\\"\\",\\"desc\\":null}"`;

exports[`RichUrl fromFile 表现正常 1`] = `"rich://{\\"url\\":\\"blob:http://localhost/ffsfeeer44\\",\\"desc\\":{\\"name\\":\\"h.css\\",\\"size\\":0,\\"type\\":\\"text/css\\",\\"lastModified\\":1574155444339}}"`;

exports[`RichUrl parse 富链接时解析正常: url不为空 1`] = `
Object {
  "url": "http://gg.gg",
}
`;

exports[`RichUrl parse 富链接时解析正常: url为空 1`] = `
Object {
  "url": "",
}
`;

exports[`RichUrl parse 富链接时解析正常: 描述为 [] 1`] = `
Object {
  "desc": Array [],
  "url": "",
}
`;

exports[`RichUrl parse 富链接时解析正常: 描述为 { x: 1 } 1`] = `
Object {
  "desc": Object {
    "x": 1,
  },
  "url": "",
}
`;

exports[`RichUrl parse 富链接时解析正常: 描述为 1 1`] = `
Object {
  "desc": 1,
  "url": "",
}
`;

exports[`RichUrl parse 富链接时解析正常: 描述为 desc 1`] = `
Object {
  "desc": "desc",
  "url": "",
}
`;

exports[`RichUrl parse 富链接时解析正常: 描述为 null 1`] = `
Object {
  "desc": null,
  "url": "",
}
`;

exports[`RichUrl toFile 文件富链接正常 1`] = `
Object {
  "lastModified": 1574155444339,
  "name": "h.css",
  "size": 0,
  "type": "text/css",
  "url": "blob:http://localhost/ffsfeeer44",
}
`;

exports[`RichUrl toFile 非文件富链接报错 1`] = `"richUrl 不是一个合法的文件富链接: \\"rich://{\\\\\\"url\\\\\\":\\\\\\"http://gg.gg\\\\\\",\\\\\\"desc\\\\\\":{\\\\\\"x\\\\\\":1}}\\""`;

exports[`RichUrl transform 表现正常 1`] = `
Object {
  "icon": "url: icon, desc: {\\"name\\":\\"icon\\"}",
  "id": 1,
  "title": "hello",
  "users": Array [
    Object {
      "avatar": "url: 2-avatar, desc: undefined",
      "id": 2,
    },
    Object {
      "avatar": "url: 3-avatar, desc: {\\"url\\":\\".\\"}",
      "id": 3,
    },
    Object {
      "albums": Array [
        "url: http://gg.gg, desc: undefined",
        "url: https://gg.gg, desc: undefined",
        "url: ftp://gg.gg, desc: {\\"ftp\\":true}",
      ],
      "avatar": "url: , desc: {\\"avatar\\":\\"4\\",\\"id\\":4}",
      "id": 4,
    },
  ],
}
`;

exports[`RichUrl transformFile 表现正常 1`] = `
Object {
  "icon": "url: blob://x0, file: icon",
  "id": 1,
  "title": "hello",
  "users": Array [
    Object {
      "avatar": "url: blob://x1, file: undefined",
      "id": 2,
    },
    Object {
      "avatar": "url: blob://x2, file: undefined",
      "id": 3,
    },
    Object {
      "albums": Array [
        "url: blob://x3, file: undefined",
        "url: blob://x4, file: undefined",
        "url: blob://x5, file: undefined",
      ],
      "avatar": "url: blob:http://localhost/ffsfeeer44, file: h.css",
      "id": 4,
    },
  ],
}
`;
